{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding the Regulatory Regions/Genes\n",
    "\n",
    "LHY, CCA1, and TOC1 are able to control the transcription of other genes because the regulatory proteins that they encode are transcription factors, or master regulatory proteins that turn other genes on and off. A transcription factor regulates a gene by binding to a specific short DNA interval called a regulatory motif, or transcription factor binding site, in the gene's upstream region, a 600-1000 nucleotide-long region preceding the start of the gene. For example, CCA1 binds to AAAAAATCT in the upstream region of many genes regulated by CCA1.\n",
    "\n",
    "The life of a bioinformatician would be easy if regulatory motifs were completely conserved, but the reality is more complex, as regulatory motifs may vary at some positions, e.g., CCA1 may instead bind to AAGAACTCT. But how can we locate these regulatory motifs without knowing what they look like in advance? We need to develop algorithms for motif finding, the problem of discovering a “hidden message” shared by a collection of strings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import import_ipynb\n",
    "from Week_1 import *\n",
    "from Week_2 import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A brute force algorithm for motif finding\n",
    "Given a collection of strings Dna and an integer d, a k-mer is a (k,d)-motif if it appears in every string from Dna with at most d mismatches. For example, the implanted 15-mer in the strings above represents a (15,4)-motif.\n",
    "\n",
    "Implanted Motif Problem: Find all (k, d)-motifs in a collection of strings.\n",
    "\n",
    "Input: A collection of strings Dna, and integers k and d.\n",
    "\n",
    "Output: All (k, d)-motifs in Dna.\n",
    "\n",
    "Brute force (also known as exhaustive search) is a general problem-solving technique that explores all possible solution candidates and checks whether each candidate solves the problem. Such algorithms require little effort to design and are guaranteed to produce a correct solution, but they may take an enormous amount of time, and the number of candidates may be too large to check."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def motif_enumeration(dna, k, d):\n",
    "    dna = \" \".join(dna)\n",
    "#     patterns =[]\n",
    "    patterns_in_all = {}\n",
    "    first_str = dna.split(\" \")[0]\n",
    "    for i in range(len(first_str)- k+1):\n",
    "        kmer = first_str[i:i+k]\n",
    "        # 1st str's kmer neighbors\n",
    "        for kmer_nbr in  neighbors(kmer, d):\n",
    "            found_in_all = True\n",
    "#             inall = {}\n",
    "            inall = []\n",
    "            # look in each dna str\n",
    "            for dna_str in dna.split(\" \"):\n",
    "                # look in each dna str for kmer neighbor with atleast d mismatches\n",
    "                found = False\n",
    "                # copy of motif found in str\n",
    "                nbr_in_str = \"\"\n",
    "                for j in range(len(dna_str)-k+1):\n",
    "                    if hamming_distance(kmer_nbr, dna_str[j:j+k]) <= d:\n",
    "                        # if found: look in next dna_str\n",
    "                        found = True\n",
    "                        # update the found motif\n",
    "                        nbr_in_str = dna_str[j:j+k]\n",
    "                        break\n",
    "                # if kmer is not found in any of the dna_strs, look for next kmer\n",
    "                if found == False:\n",
    "                    found_in_all = False\n",
    "                    break\n",
    "                # add the\n",
    "                else:\n",
    "                    # if motif is found: add its version in a list\n",
    "                    inall.append(nbr_in_str)\n",
    "            if found_in_all == True:\n",
    "                # if motif confimed: add it to lists\n",
    "#                 patterns.append(kmer_nbr)\n",
    "                patterns_in_all[kmer_nbr] = inall\n",
    "    return patterns_in_all\n",
    "#     return \" \".join(list(patterns_in_all.keys()))\n",
    "#     return list(patterns_in_all.keys())\n",
    "a = \"\"\"ATTTGGC TGCCTTA CGGTATC GAAAATT\"\"\".split(\" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given a k-mer Pattern and a set of strings Dna = {Dna1, … , Dnat}, we define d(Pattern, Dna) as the sum of distances between Pattern and all strings in Dna,\n",
    "\n",
    "d(\\textit{Pattern},\\textit{Dna}) = \\sum\\limits_{i=1}^t d(\\textit{Pattern},\\textit{Dna}_i).\n",
    "d(Pattern,Dna)= \n",
    "i=1\n",
    "∑\n",
    "t\n",
    "​\t\n",
    " d(Pattern,Dna \n",
    "i\n",
    "​\t\n",
    " )."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def consensus(seqs):\n",
    "    cons_seq = \"\"\n",
    "    for i in range(len(seqs[0])):\n",
    "        count = {\"A\":0, \"C\":0, \"G\":0,\"T\":0}\n",
    "        for seq in seqs:\n",
    "            count[seq[i]] +=1\n",
    "        count_max = max(count.values())\n",
    "        for key,value in count.items():\n",
    "            if value == count_max:\n",
    "                cons_seq += key\n",
    "                break\n",
    "    return cons_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns miniumum sum of distances between pattern and string in DNA (using scoring: Hamming Distance)\n",
    "def d(pattern, dna):\n",
    "    k = len(pattern)\n",
    "    distance = 0\n",
    "    motifs = []\n",
    "    for dna_str in dna:\n",
    "        str_dist = k\n",
    "        motif_str = \"\"\n",
    "        for i in range(len(dna_str)-k+1):\n",
    "            if str_dist >= hamming_distance(pattern, dna_str[i:i+k]):\n",
    "                str_dist = hamming_distance(pattern, dna_str[i:i+k])\n",
    "                motif_str = dna_str[i:i+k]\n",
    "        distance += str_dist\n",
    "        motifs.append(motif_str)\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Median String Problem\n",
    "Our goal is to find a k-mer Pattern that minimizes d(Pattern, Dna) over all k-mers Pattern, the same task that the Equivalent Motif Finding Problem is trying to achieve. We call such a k-mer a median string for Dna.\n",
    "\n",
    "Median String Problem: Find a median string.\n",
    "\n",
    "Input: A collection of strings Dna and an integer k.\n",
    "Output: A k-mer Pattern that minimizes d(Pattern, Dna) among all possible choices of k-mers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def median_string(dna, k):\n",
    "    if dna[0].islower():\n",
    "        for i in range(len(dna)):\n",
    "            up = dna[i].upper()\n",
    "            dna[i] = up\n",
    "    distance = len(dna[0])\n",
    "    median = \"\"\n",
    "    for i in get_all_kmers(k):\n",
    "        if distance > d(i, dna):\n",
    "            distance = d(i, dna)\n",
    "            median = i\n",
    "    return median\n",
    "\n",
    "\n",
    "\n",
    "mtf = \"\"\"CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC\n",
    "GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC\n",
    "GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG\n",
    "\"\"\".split(\"\\n\")\n",
    "\n",
    "# median_string(mtf, 7)\n",
    "# mtf\n",
    "# d(mtf, \"AATCCTA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_gen(motifs):\n",
    "    profile = []\n",
    "    for pos in range(len(motifs[0])):\n",
    "#         nc_count = [0 for i in range(4)]\n",
    "        # as per laplase law\n",
    "        nc_count = [1 for i in range(4)]\n",
    "\n",
    "        for motif in motifs:\n",
    "            if motif[pos] == \"A\":\n",
    "                nc_count[0] +=1\n",
    "            elif motif[pos] == \"C\":\n",
    "                nc_count[1] +=1\n",
    "            elif motif[pos] == \"G\":\n",
    "                nc_count[2] +=1\n",
    "            else:\n",
    "                nc_count[3] +=1\n",
    "        nc_prob = [round(i/sum(nc_count), 3) for i in nc_count]\n",
    "        profile.append(nc_prob)\n",
    "    # to transpose the profile use this code\n",
    "    profile_t = []\n",
    "    for i in zip(*profile):\n",
    "        profile_t.append(list(i))\n",
    "    profile = profile_t\n",
    "    return profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_most_probable(dna, k, pr):\n",
    "    scores = []\n",
    "    nc_list = {\"A\":0, \"C\":1, \"G\":2 ,\"T\":3}\n",
    "    for i in range(len(dna) - k + 1):\n",
    "        score = 1\n",
    "        for pos in range(k):\n",
    "            nc = dna[i:i+k][pos]\n",
    "            score *= pr[nc_list[nc]][pos]\n",
    "        scores.append(score)\n",
    "    maxscore = max(scores)\n",
    "    return dna[scores.index(maxscore):scores.index(maxscore)+k]\n",
    "\n",
    "\n",
    "prfl = \"\"\"0.4 0.3 0.0 0.1 0.0 0.9\n",
    "0.2 0.3 0.0 0.4 0.0 0.1\n",
    "0.1 0.3 1.0 0.1 0.5 0.0\n",
    "0.3 0.1 0.0 0.4 0.5 0.0\"\"\".split(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Greedy Motif Search\n",
    "Our proposed greedy motif search algorithm, GreedyMotifSearch, starts by forming a motif matrix from arbitrarily selected k-mers in each string from Dna (which in our specific implementation is the first k-mer in each string). It then attempts to improve this initial motif matrix by trying each of the k-mers in Dna1 as the first motif. For a given choice of k-mer Motif1 in Dna1, it builds a profile matrix Profile for this lone k-mer, and sets Motif2 equal to the Profile-most probable k-mer in Dna2. It then iterates by updating Profile as the profile matrix formed from Motif1 and Motif2, and sets Motif3 equal to the Profile-most probable k-mer in Dna3. In general, after finding i − 1 k-mers Motifs in the first i − 1 strings of Dna, GreedyMotifSearch constructs Profile(Motifs) and selects the Profile-most probable k-mer from Dnai based on this profile matrix. After obtaining a k-mer from each string to obtain a collection Motifs, GreedyMotifSearch tests to see whether Motifs outscores the current best scoring collection of motifs and then moves Motif1 one symbol over in Dna1, beginning the entire process of generating Motifs again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def score(motifs):\n",
    "    total_score = 0\n",
    "    for pos in range(len(motifs[0])):\n",
    "        nc_count = [0 for _ in range(4)]\n",
    "        for motif in motifs:\n",
    "            if motif[pos] == \"A\":\n",
    "                nc_count[0] +=1\n",
    "            elif motif[pos] == \"C\":\n",
    "                nc_count[1] +=1\n",
    "            elif motif[pos] == \"G\":\n",
    "                nc_count[2] +=1\n",
    "            else:\n",
    "                nc_count[3] +=1\n",
    "        total_score += sum(nc_count) - max(nc_count)\n",
    "    return total_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 CCGGCGGG 12\n",
      "9 CCATCGGCC 17\n",
      "10 CTATCGGCCC 19\n",
      "11 GGACTTCCGGC 20\n",
      "12 GGACTTCCGGCC 24\n"
     ]
    }
   ],
   "source": [
    "def greedy_motif_search(dna, k, t):\n",
    "    best_motifs = [dna_str[:k] for dna_str in dna]\n",
    "    for i in range(len(dna[0])-k+1):\n",
    "        # move through first dna string\n",
    "        motifs = [dna[0][i:i+k]]\n",
    "        # make profile of motifs (to be updated)\n",
    "        motif_profile = profile_gen(motifs)\n",
    "        for j in range(1,t):\n",
    "            # improve profile by adding more motifs by finding most_probable motifs in next dna strings\n",
    "            motifs.append(profile_most_probable(dna[j], k , motif_profile))\n",
    "            # update the profile\n",
    "            motif_profile = profile_gen(motifs)\n",
    "        if score(motifs) < score(best_motifs):\n",
    "            best_motifs = motifs\n",
    "    return best_motifs\n",
    "\n",
    "m = \"\"\"ttACCTtaac\n",
    "gATGTctgtc\n",
    "acgGCGTtag\n",
    "ccctaACGAg\n",
    "cgtcagAGGT\"\"\".upper().split(\"\\n\")\n",
    "# m = \"\"\"GGCGTTCAGGCA\n",
    "# AAGAATCAGTCA\n",
    "# CAAGGAGTTCGC\n",
    "# CACGTCAATCAC\n",
    "# CAATAATATTCG\"\"\".split(\"\\n\")\n",
    "mot = \"\"\"GGCGTTCAGGCA\n",
    "AAGAATCAGTCA\n",
    "CAAGGAGTTCGC\n",
    "CACGTCAATCAC\n",
    "CAATAATATTCG\"\"\".split(\"\\n\")\n",
    "\n",
    "mot = \"\"\"TATTCACCCTTTAGGCTTACTGCTCGGAGTCAGAAGGCGAGACGCAACATCTTTAGCTATCGTGGTAAGGTATTGTTGCTACTAGCACCGGTGCTACCAGCCCGGTGTCAGCACCTGTGCTATTTACCCCCTATAGCTGCGACGTCTAATTAGCAG\n",
    "CGTCAGGCGGGCCTACGGTATCCATATAAGCAGTAATTCGATTATCTTTGATTTGCCGCTTAGGATTCGAGTATAGATGCGAGGACTCCTCTTGCTCCGTCAGATATTTAAACTGGAAAACGTATCCTCGGATATTTATCAGATGACATTGTTCGT\n",
    "TCCGGTAGCTAGTGTGAGGCCTGCTAAGGCGTCGAGTACCATCCAAACGTAGATGTGCTACCTCGCTCCATACAACTCGGGCATACTACCGATACTCTATCAAGAAAGGGCTGTTGCCCGGATTTAAATGACATACCTGCGACTAGGACCTCGCCT\n",
    "AGACTGACCGATCATGGAGAGAAGTATCGACATTTTTCGCTGACCCTCAGGGGAGGGAGCAAATCCATCGCGAGAGCCCGTCGCGATGTCGTGAGCGCTAGCTTCACGGTGTGGGAACTGGGCGCAGTGACTATATTTGCGACCTCCCTTATAGGG\n",
    "CCGGACGTAACTGTTGGGCTTGGAGCGGCTTGTCGGGAGTTGCGCTGAGATGCCATTGCCATCCAGGATCTAATGTATTAATGCACTTACATGCCCGCGTTTTAGTCTTAACGATGGACGGCTACAATTCACATACTTGCGATCAGGATTAGGATT\n",
    "TATTTTCCCCACAATGTGAATCGGATAGATGCGACCGGACTTACAGCGTCCCATTGCAATCTTGATATGGGGTTTAGAGATCTATGTTCATTATGCTTTCTCATGGGAGTAAAGCATGGATACTCGTAATCTCAAGTAAATGCTAAGTCGCATTAC\n",
    "ACTTGCGTCAGGTCGTTGATATCATCTCATCCTGCCAATACATAACTACATCAGCCACCCACAGGTAATAAGGCTAGCTCGTCAAAACAACAGCTGTAAAATCGGATTGAGTGTGAAGTAAGAGAGGGTGACGAGCTATCGTCGATACGTGCGAAC\n",
    "GAAGAGGGCTTCGCCAGCTGAACTATAGTTGCGACTACTGAAAGCGCAAAGATCCGATATAAGCCAAAGAAAGCCTTGATCTCCTTTGCCCTGCCTCACCGGACTCCACACTCATCTGAACGGGAGGTCAGTCTAAGGAAGCTACCCTGTTCATTC\n",
    "GTCTATCTGCCGGGGTTGAGTTACATGCTCGCGGTATCCAGGACCGAAATACATGCGAAACGAGCCGCTCGGGTCCTTGGGCCAGTCGGCTCCTACGTAATGAACCTATCCGCACCTACTCGTAGGCTTACATTGGTTTTTTATCTTATCGTAGTT\n",
    "AGGGGATACCTCTGGGGGAATGGGCTTATTGGGTTTGGCATGACTCATCATTGTTCCCCGTAATCCCTGGGAGGGCTTCCCCAATGCCGCTAGTTCTCCGCTGAGTTCTTCTGCCGGTACGCAACGCGTAGTATAGCTGCGACTCCATGATGATCC\n",
    "TAGAAATACCCGGAGCAGGGCCAGGGCGGTTACCTCACAGATAGAGCTGCTGCAGCGCTGATCCAAGACAAGATAGATGCGAAGCGTATCAACCTCCTGCTTGGGGGACGATTTCCTCCGCTGCGACTCAGTTAGCTTGGTAGCAACATCCTACAA\n",
    "CCCTAGGATTCGATAACTGCGATCGCCGAGCTTTACATTAAGTACTCGCCTCTACATTTAAAGCGAGGTGATAAGGACAGCTAGCATCCGAAGAATACAGTTATTCTCAACAACCACCACTTCATTCCTACTCGTACGTCGTGCCGTACTTTTTAT\n",
    "TATGCGTCGCCTCGATACATGCCCGGAAGGACGTTCACACACTAAAAATTGTGCAAGGTAACGATCATGCGCGGACTTCGTGTTAACCGACAAGGCCAGGAATAAAGTCTGAATATCGTAATAGATGCGATGCAGTGCGCGAAACTACCTCTCGTG\n",
    "CGTGAACCTGCCAAGTTAAGCTACGCGCCGCATGACTCGAACACAATAATAGATATTGAGGAGCTTAAAGGGCAGACCGATAGGTGACGGTGGTTTCAGGTGGCTCACCTAGTTCGTACAACTGCTGGCCTGATAAATGCGAAGAATGGTCACGGT\n",
    "TGTGGAAATGGTCTGGTACGCTTGACGCTATGCTCGTCGTAGTTGCATCCACATATAGAGATACGTGCGATGCCTTCGACCGTTAAGGGGCATTCTCGCCACGCAACTCCATGTTTCTACAAAGCCGTTCATGTTAACAAGTTGGTTTTCTGTACG\n",
    "AGAAGTACTCCATAACCATTTCGGGTATTAGAAAGCATACCTGCGAGGACATGCTGGTGGTCATAAATAGGGGACTACGTACGACGAGATAGTTGTCTTTCCGCAAACACGGGTCACTGAGGACCCCCGTCGGATTCTGACCGACATTCTATTTTT\n",
    "CTATTGCCCTCTTTGCCACCAGCCGATACTAGGATTCACGTAACGCTAACGTGGCGGGCGATACTTGCGACCGTGCACGCCTACGAGTTAGTCTAGTTGGGGCGTTAGCTGACATCAACAGTGTCCATGGGGTTTCCCCGACTCTGGCGCGGCTAA\n",
    "TTTATTCTAGGTGAACCGAGGCGCATATATGCGACTCTGGGACACACAAAACTGCAAGAGCTGGCGATCTATGGGACCCTGTACGCTAACCCGGTTACACTGTCGACCATACAGCAATGGTAAGTCTAAAGGGAGGGGCTGAGCGCGAAATGTCCA\n",
    "GTCATGCTGGGCGCATTGAGTGAAATAGCTGCGACTTGGCGACGTCACGAGACGCGGTATGTGTTGAAGGCAACTGGCTGTCAGACTTTTATGAATTGGAACGAGGAACCCTGTAGCCATATACAGGTCACGATATGTGCGGTCCTAACTTCCTCC\n",
    "CACGACTCACCCTGACTAGGTCATAGATGCCACCCGCGTTCTACGAAAGCCCTCCGGGTGGCAAATGCATTGACCTTCCCTGCTCTCTCCAAAAGTAGTTTCGAGAAGCTAGATTGATTGCTGTATAGGGTTATAAGTGCGAGTAGGCATCATGAG\n",
    "GGTCCACGTTGTAACCAGCTGGTAGTTTTATAATTTGGAAGCCTAATCCTGCGCGCACAGCAACAGTTAGATCAAGCCTAGTGTAACACCTGGTTGCGCATGGCTGTGTGTAGTAACTTGAAAGCATGGAACGACAAAAGTAGTATACATGCGAGA\n",
    "CCCACACCATCACAGCGGTGCGGTGACAATGTTTCAACTACCTTCCTCCACGGTAGGAAGGGGCAGATACAGAGTTACCGGCGGACTATCCGCGGAACCCACAAACGACATACGCCGCGTTCAGTGGAGCGGTCTGCTTGGGTCATACCTGCGATA\n",
    "CCGCATCGGTAGGTTCAACCCCCGAAGTATGACAAAACATTTAGGCGCATAGATGCGAGGCACAATAAGAGCGCATTGCTCAAGCAGACTACCTGAACACTACAGTAATATGAAAATCGTGCAAAGTCTGTCAGGTTAGGCCATTGATTAACGGTC\n",
    "ATACTATGGCACCTCGGAGCTGCGTCCGAATTGCAAGATCCTGGTTAACAGAAGTGCAACTAATTCTGGCACGTATCATACCGTGTTTCCAGGCCCATAGAAGCTGATATATTTGCGAAGCAACCTTCGCGAACAGGGATGCCTCTTCGTCGTACT\n",
    "TCCAAATGGTATCTTCAGCGGATTACGTGACGTCCTGATCTTGTGTGAGCGTTAGCCGTGCATCCAGCATACTCGGAACGTATGTCGTCGACTCTGTCTCAGTACGAACGCATATTCGTCCATGAGATCACGACTCGCTCCCGAATAGATGCGAGA\"\"\".split(\"\\n\")\n",
    "mot =\"\"\"TGAGATAGCGGATGGTCTGTCGCCTTCTTGTCTGATACCGCCTTCTTCTACCGACAGAAACTAAGGCCCAGCGACATGCTCAGTCAAATTGGGTCAAATTTCGTGGTATAGTTGGAAGACTAGATTTGGGTGCGTCAATCGATGAGCCAACCTCGC\n",
    "GGTCGGTATTTGCCTTTTGATGTAGACTATATGTGGTCTCAACTATAGTTATCACGGTGTGGCGGCCTGTGGCCAAGGGCCACATGAGGTCAGGTTTTACAGTCCACGACCGCCTTCGGTACTGTGAACAGTAGGAATCGGTCATGGTAGGTTTAT\n",
    "GGGCTTGTACCGCGCGGTTGTTCTCCGATTCGTGAGCCCGTCCGGAGCTATGTATGTTCGTGGGTATCTCAACCGCGCGCCGGCTGCACCCCTTTCCGAAGGGCCAGGACGGTGACGATTTGAGTATGAGGGATGTAACCCTATCAATGACTGTCC\n",
    "CGAAGGACCAGATCACCTACGGGCAGGAGAGATCGCAACTAAGGGAAGGTCGCGTGCTGCGCGCCAGAATACATACGCGAGCTCCCGATGGCACATGGATTGGTAAGCCTAGATCATCGGTCGTGCAGAGCCGAATACGGCCTGTCCGCACTGTTA\n",
    "ATATGCCCTAAATTGAATATACTTGGCGCCTTACTGCCAAGGTCCACTGGCGGGCGGTTTCTCGGATAGGGGCACAATGTCGAGCAGGCCATAAGCTGGCGTTGCAACGTACAAGAAGATGTGGAACTCACACAAACTTACAATAGCGCGTGAGCA\n",
    "CGCGAGGCGATGGGGGGGATCATCGCTGAATGCGCCGGCATACGAAGGTATCTATGCAAATATAACCATAAATCACTCGATGTGGCTAGGCCGCTGACTGCGGTACACTGTTCCGCGACCCAAAGGCCCACCCGGTCAGTCGTTCTCGGAGTATGT\n",
    "TTCCGTGGGTGTCTATCGTTGGCTCAAAGGCCCACTCGTTACAGCCCATTAGAATAGTCCCATGGTGGCAGGCGCTTCTATGGCGCTATCGCCCAGCAGTCCTTTGCGAACTGGAGACAATGAGGACTCGCGTCAGGGTACACGTATGCGACCGCT\n",
    "CAGTAATGGCGACTAAGGGCCATCATGGTCAGGCGCATGGAATGTTGGTCATTTTACTTTTGATTGTATGACCCGTAACCGTGACAAAGGCTCGGACACCATACAATGTGGAGGAGAAATCACTTCTACAGCGGATGGGCGGGTTTTAAAAGAGAC\n",
    "GTCCCGCATGCACAAAGGACCAAGCCAGTACGTTCGAGGGTCATCTATATGAACTTGACAGTCTTTACCGCAGGTTACCGTTCGAACTGGGGTTTGAAGTTTTTCTGACGTAATAAACACTTCTCGGTAGCGTAGGTCTCCAGACACCGCCGCTGT\n",
    "TACGTTGTCAAGACATAAGAACCTAGTGATTATTAAACCGCTGGCCAAACCGAAGCGCCCTCCCGATCCTCTGTCTCCCATGGTGCTTTCATTAAGCGTCAAGCCCAACGAAGGGCCACGTCGTCTTTCGATGCGCGTGGAATGAGTTGGAGCCTG\n",
    "GATCATTGCAGCGACTATCAGCTTAATCTTCTCGACAGGACACCATCGCAGATTCTCCGACCCTCGGCCGCTGAAGCAACCAACGCTGTAGCTGCAGTAGAGTGGAGAGGTGGCGAATGGGTGTCTCAGCACCAAAGGTCCAAAGAACTAGCTCAA\n",
    "ATAAACGAGATGCTCGCAAGTGGAACTCCATCCTATCGAAGGTCCACCAGAACTGCAGAGGCCGACATGCTATTCCTCGCACATTGCAAAGTATGATGGAAAAGAATGAGTCGAGAGTAAACAAGTAGCGATATATTCTGCTAACGTCCCCTGGGG\n",
    "GTTGCTGATTATAGGCCTGGGAAGCCAAGGGCCACGTAAATGATCACTGTTGTACCGGATGTGTAATATTCGTTACGGACACTGTTAGCTGATACACCGCAAGTAATAGTGTCTGGTACCATCGTGAGTGGGATATACTTGATGTTCCAAAGTATT\n",
    "AGGAGCACCGGTCAAAGGACCATTTCGTTCGAATGGGAAACACTTAGGGCCAGGGAGATTCTCGGACTAGTGCAAGCACTTCTTTGCCTTAGCACAGACAGATTTATACCGGATGACGTGTTTGACTTCAAAAAACTTCCTGAAGTCGACAGGGTA\n",
    "GTATCAACATCGCTAAGGGCCACTTATGACTTCAAGCATCGCAATCTCGTTCCGAGCTCGCTGAGCTCTCCTATGTTGCCTCCACAACTTGGTAAAGTTCCGGTTGTCCTGGCTATATTTGGTACCTTTCGAATAACAATCGCAAAATCTATTTAT\n",
    "TGTAATTTCGTCCGAAGGCCCAGCTGTATACTTCCCCTGCGGACCATGATCAGGCAAGTGGCGATACTTTAATTCTCTCACGGGTTAGTGCTGGGCGCTAGTTGTTACGTCCAGGAACATCCTACGCAGGAAGATCAGTAGCTCCATTTCAGGCGC\n",
    "AATCGGACATTTAACGCCAACTTGTTGCCATTTCAGCTCTCCCCAAGAAACCGCAGCTACTACGAGCCTCGGCAAAGGTCCACCCCCTCAAACATTTTCTCTGGCAAACGATCCATTAGCCAGAAGAGGTACAAAAAAGACAAACTCTCACGATCG\n",
    "ATTCTTCTGGATTCTGTGAAGGAATGTCATATTTACCACTCGAATACACTGAGGTACACCCTGTAAGAATGACAAAGGGCCAGTCTTTGGAAGGGAGATGATGGTTCCAATAGTCAGACCTCAGTTACAGTGAGGCTCGAACCTCACCACGCTCAG\n",
    "AGTCCTGTGCACCTAAGGGCCAGATGGCCTGGAGCCTTCTTTGGGCCGCTTTCCGGCGCGATCTGGGTGGCTTGAGGGGGTTCACTTCTAAGCTGACGGTAACTATTATATAAACATGCACTAGTGACATTCAGCGTTTCACTTAACCCCATTCGA\n",
    "CCGGAGCGGGAGCAGACTCGGTTCTCAACATGGGAGGAAAATCCGCCATCTATCGTGCTGGCTCGCCGAATGCTGGGTCCCATAGCTAAGGAGGATAGGACGTGCTAGAATCTCTCGGGACTAAGGACCATGCAAGTGCCGGCAAATGGCGTGCAG\n",
    "AGAGGCTCTACACGAGAAGCGGAGCGGCTCTAGGCCGTGCAGGGATTGCGGACTTATGGGCTGTAAAGTAAGTTAGCGCGTGCAGGTTCACAAGTCCTAAGGTCCACACTACAGAATCTCAATGAGCGTGGCTCATATCAGTCCGGTGGCTGACAT\n",
    "ACAACGCGAGGAAGGTCCGTATCCTGGCGATGATGTGTATGGCCATGAGGCCAACGAGACCTTAAAGTGCGTCACACCGCCAAAGGGACCAACCTGCCGGCATGGAAAATACACATTTGCCTAAGGACCAACGCTGGCTATGTATCGCGGGCGCGT\n",
    "TTTGAAACACCTCCTCAGTCCATCCATATGATGCTGTGGTCGTATCGACCGTCTAAAAGGGAGGCAAGAGTGCCAAGGTCCATTTACAATGTTCACCGTTACAGCCAAATCGTCTTTACTCCGTTGACATCTAAGCTGGTGTCTTCGACGCCAACT\n",
    "TTAGGTCATTTGAAGCTCAGGCAGGCTCCCATTGAGATCAGCAAAGCTTCAGTGTTAAAACGTTCCCTCCAGCTGTTCTTAAGATCAGCACCTGTCCGAAGGCCCAGATTACACGCCGACTCACGCTCAGAAAGCGAGCGTTTCTTATTTACATGC\n",
    "CGCGCCACAGCGTAATGATCGTAAGCGATAATACCCTCTTGGGAGCGAGGTAAGTTTATGGGTATCCTCGCGGGTTTCGCCGATCTGCCTTTTTTGCTAGCGCACTCCTCCCCTAATCCTCAAAGGTCCAAAGAAATTATGGACACTATCAAGAGG\"\"\".split(\"\\n\")\n",
    "\n",
    "dnaa = \"\"\"GCGCCCCGCCCGGACAGCCATGCGCTAACCCTGGCTTCGATGGCGCCGGCTCAGTTAGGGCCGGAAGTCCCCAATGTGGCAGACCTTTCGCCCCTGGCGGACGAATGACCCCAGTGGCCGGGACTTCAGGCCCTATCGGAGGGCTCCGGCGCGGTGGTCGGATTTGTCTGTGGAGGTTACACCCCAATCGCAAGGATGCATTATGACCAGCGAGCTGAGCCTGGTCGCCACTGGAAAGGGGAGCAACATC\n",
    "CCGATCGGCATCACTATCGGTCCTGCGGCCGCCCATAGCGCTATATCCGGCTGGTGAAATCAATTGACAACCTTCGACTTTGAGGTGGCCTACGGCGAGGACAAGCCAGGCAAGCCAGCTGCCTCAACGCGCGCCAGTACGGGTCCATCGACCCGCGGCCCACGGGTCAAACGACCCTAGTGTTCGCTACGACGTGGTCGTACCTTCGGCAGCAGATCAGCAATAGCACCCCGACTCGAGGAGGATCCCG\n",
    "ACCGTCGATGTGCCCGGTCGCGCCGCGTCCACCTCGGTCATCGACCCCACGATGAGGACGCCATCGGCCGCGACCAAGCCCCGTGAAACTCTGACGGCGTGCTGGCCGGGCTGCGGCACCTGATCACCTTAGGGCACTTGGGCCACCACAACGGGCCGCCGGTCTCGACAGTGGCCACCACCACACAGGTGACTTCCGGCGGGACGTAAGTCCCTAACGCGTCGTTCCGCACGCGGTTAGCTTTGCTGCC\n",
    "GGGTCAGGTATATTTATCGCACACTTGGGCACATGACACACAAGCGCCAGAATCCCGGACCGAACCGAGCACCGTGGGTGGGCAGCCTCCATACAGCGATGACCTGATCGATCATCGGCCAGGGCGCCGGGCTTCCAACCGTGGCCGTCTCAGTACCCAGCCTCATTGACCCTTCGACGCATCCACTGCGCGTAAGTCGGCTCAACCCTTTCAAACCGCTGGATTACCGACCGCAGAAAGGGGGCAGGAC\n",
    "GTAGGTCAAACCGGGTGTACATACCCGCTCAATCGCCCAGCACTTCGGGCAGATCACCGGGTTTCCCCGGTATCACCAATACTGCCACCAAACACAGCAGGCGGGAAGGGGCGAAAGTCCCTTATCCGACAATAAAACTTCGCTTGTTCGACGCCCGGTTCACCCGATATGCACGGCGCCCAGCCATTCGTGACCGACGTCCCCAGCCCCAAGGCCGAACGACCCTAGGAGCCACGAGCAATTCACAGCG\n",
    "CCGCTGGCGACGCTGTTCGCCGGCAGCGTGCGTGACGACTTCGAGCTGCCCGACTACACCTGGTGACCACCGCCGACGGGCACCTCTCCGCCAGGTAGGCACGGTTTGTCGCCGGCAATGTGACCTTTGGGCGCGGTCTTGAGGACCTTCGGCCCCACCCACGAGGCCGCCGCCGGCCGATCGTATGACGTGCAATGTACGCCATAGGGTGCGTGTTACGGCGATTACCTGAAGGCGGCGGTGGTCCGGA\n",
    "GGCCAACTGCACCGCGCTCTTGATGACATCGGTGGTCACCATGGTGTCCGGCATGATCAACCTCCGCTGTTCGATATCACCCCGATCTTTCTGAACGGCGGTTGGCAGACAACAGGGTCAATGGTCCCCAAGTGGATCACCGACGGGCGCGGACAAATGGCCCGCGCTTCGGGGACTTCTGTCCCTAGCCCTGGCCACGATGGGCTGGTCGGATCAAAGGCATCCGTTTCCATCGATTAGGAGGCATCAA\n",
    "GTACATGTCCAGAGCGAGCCTCAGCTTCTGCGCAGCGACGGAAACTGCCACACTCAAAGCCTACTGGGCGCACGTGTGGCAACGAGTCGATCCACACGAAATGCCGCCGTTGGGCCGCGGACTAGCCGAATTTTCCGGGTGGTGACACAGCCCACATTTGGCATGGGACTTTCGGCCCTGTCCGCGTCCGTGTCGGCCAGACAAGCTTTGGGCATTGGCCACAATCGGGCCACAATCGAAAGCCGAGCAG\n",
    "GGCAGCTGTCGGCAACTGTAAGCCATTTCTGGGACTTTGCTGTGAAAAGCTGGGCGATGGTTGTGGACCTGGACGAGCCACCCGTGCGATAGGTGAGATTCATTCTCGCCCTGACGGGTTGCGTCTGTCATCGGTCGATAAGGACTAACGGCCCTCAGGTGGGGACCAACGCCCCTGGGAGATAGCGGTCCCCGCCAGTAACGTACCGCTGAACCGACGGGATGTATCCGCCCCAGCGAAGGAGACGGCG\n",
    "TCAGCACCATGACCGCCTGGCCACCAATCGCCCGTAACAAGCGGGACGTCCGCGACGACGCGTGCGCTAGCGCCGTGGCGGTGACAACGACCAGATATGGTCCGAGCACGCGGGCGAACCTCGTGTTCTGGCCTCGGCCAGTTGTGTAGAGCTCATCGCTGTCATCGAGCGATATCCGACCACTGATCCAAGTCGGGGGCTCTGGGGACCGAAGTCCCCGGGCTCGGAGCTATCGGACCTCACGATCACC\"\"\".split(\"\\n\")\n",
    "\n",
    "\n",
    "for k in range(8,13):\n",
    "    ans = greedy_motif_search(dnaa,k,10)\n",
    "#     print(ans)\n",
    "    print(k ,consensus(ans), score(ans))\n",
    "# consensus(['ACCT', 'ATGT', 'ACGG', 'ACGA', 'AGGT'])\n",
    "# consensus()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input: DNA Strings and a profile matrix\n",
    "### Output: Most probable motifs in all the DNA-strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ACCT', 'ATGT', 'GCGT', 'ACGA', 'AGGT']"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_motifs(dna, profile):\n",
    "    motifs = []\n",
    "    for dna_str in dna:\n",
    "        motif = profile_most_probable(dna_str, len(profile[0]), profile)\n",
    "        motifs.append(motif)\n",
    "    return motifs\n",
    "\n",
    "# prfl = \"\"\"0.4 0.3 0.0 0.1 0.0 0.9\n",
    "# 0.2 0.3 0.0 0.4 0.0 0.1\n",
    "# 0.1 0.3 1.0 0.1 0.5 0.0\n",
    "# 0.3 0.1 0.0 0.4 0.5 0.0\"\"\".split(\"\\n\")\n",
    "# prfl = [i.split(\" \") for i in prfl] \n",
    "# prfl1 = []\n",
    "# for i in prfl:\n",
    "#     l = []\n",
    "#     for j in i:\n",
    "#         l.append(float(j))\n",
    "#     prfl1.append(l)\n",
    "\n",
    "dnas = [\"ttaccttaac\", \"gatgtcgtc\", \"acggcgttag\",\"ccctaacgag\",\"cgtcagaggt\"]\n",
    "dnas = [i.upper() for i in dnas]\n",
    "p = [[0.8, 0 ,0 ,0.2],[0, 0.6,0.2,0],[0.2,0.2,0.8,0],[0,0.2,0,0.8]]\n",
    "\n",
    "get_motifs(dnas, p)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
